libsolidity/formal/SMTChecker.cpp

```diff
@@ -828,3 +870,89 @@ smt::Expression SMTChecker::var(Declaration const& _decl)
 	solAssert(m_variables.count(&_decl), "");
 	return m_variables.at(&_decl);
 }
+
+void SMTChecker::popPathCondition()
+{
+	solAssert(m_pathConditions.size() > 0, "Cannot pop path condition, empty.");
+	m_pathConditions.pop_back();
+}
+
+void SMTChecker::pushPathCondition(smt::Expression const& _e)
+{
+	m_pathConditions.push_back(currentPathConditions() && _e);
+}
+
+smt::Expression SMTChecker::currentPathConditions()
+{
+	if (m_pathConditions.size() == 0)
+		return smt::Expression(true);
+	return m_pathConditions.back();
+}
+
+void SMTChecker::solveMinMax(BinaryOperation const& _op)
+{
+	if (m_currentLoopCondition != nullptr)
+	{
+		string warnMessage;
+		Token::Value op = _op.getOperator();
+		if (op >= Token::LessThan && op <= Token::GreaterThanOrEqual)
+		{
+			smt::Expression left(expr(_op.leftExpression()));
+			smt::Expression right(expr(_op.rightExpression()));
+			smt::Expression toMinMax = (
+				op == Token::LessThan ? (right - left) :
+				op == Token::LessThanOrEqual ? (right - left + 1) :
+				op == Token::GreaterThan ? (left - right) :
+				/*op == Token::GreaterThanOrEqual*/ (left - right + 1)
+			);
+			m_interface->push();
+			m_interface->addAssertion(expr(_op));
+			m_maximizationResult = m_interface->maximize(toMinMax);
```

Comment:
What happens if this is used in something like `while (f(a < b)) { ... }`, where the `f` ignores its argument and just depends on side-effects?

Reply:
That's one of the things I wanted to discuss. For the beginning I wanted to support only binary comparison operations (except `=` and `!=`) as the condition. I followed from `m_currentFunction` the usage of class members to keep track of the current state with the new `m_currentLoopCondition`, but it indeed doesn't work in the case you mentioned. So in order to support exactly what I wanted at first, the check would have to be "is the condition one of the supported operators?" in the scope of the loop, which looks like a hack in the pattern. So I'm still not sure about the best way to handle that...

Reply:
Yeah, I think you should check in the loop whether it conforms to the pattern. lots of `dynamic_casts`s involved, but I don't have a better solution for now.

Reply:
Gonna go with that then, thanks.